File: app/src/main/java/com/alvinhkh/buseta/service/EtaWidgetAlarm.java
package com.alvinhkh.buseta.service;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import androidx.annotation.NonNull;

import com.alvinhkh.buseta.C;
import com.alvinhkh.buseta.provider.EtaWidgetProvider;

import java.util.Calendar;

public class EtaWidgetAlarm {

    private static final Integer ALARM_ID = 0;

    public static void start(@NonNull Context context, int seconds) {
        if (seconds <= 0) return;
        Integer intervalMillis = seconds * 1000;
        Calendar calendar = Calendar.getInstance();
        calendar.add(Calendar.MILLISECOND, intervalMillis);

        Intent alarmIntent = new Intent(context, EtaWidgetProvider.class);
        alarmIntent.setAction(C.EXTRA.WIDGET_UPDATE);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, ALARM_ID, alarmIntent,
                PendingIntent.FLAG_CANCEL_CURRENT);

        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        // RTC does not wake the device up
        if (alarmManager != null) {
            alarmManager.setRepeating(AlarmManager.RTC, calendar.getTimeInMillis(), intervalMillis, pendingIntent);
        }
    }

    public static void stop(@NonNull Context context) {
        Intent alarmIntent = new Intent(context, EtaWidgetProvider.class);
        alarmIntent.setAction(C.EXTRA.WIDGET_UPDATE);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, ALARM_ID, alarmIntent,
                PendingIntent.FLAG_CANCEL_CURRENT);

        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        if (alarmManager != null) {
            alarmManager.cancel(pendingIntent);
        }
    }
}

File: app/src/main/java/com/alvinhkh/buseta/service/EtaWidgetAlarm.kt
package com.alvinhkh.buseta.service

import android.app.AlarmManager
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent

import com.alvinhkh.buseta.C
import com.alvinhkh.buseta.appwidget.FollowWidgetProvider

import java.util.Calendar

object EtaWidgetAlarm {

    private const val ALARM_ID = 0

    fun startAlarm(context: Context, seconds: Int, widgetId: Int, groupId: String) {
        if (seconds <= 0) return
        val intervalMillis = seconds * 1000
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.MILLISECOND, intervalMillis)

        val alarmIntent = Intent(context, FollowWidgetProvider::class.java)
        alarmIntent.action = C.ACTION.WIDGET_UPDATE
        alarmIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
        alarmIntent.putExtra(C.EXTRA.GROUP_ID, groupId)
        val pendingIntent = PendingIntent.getBroadcast(context, ALARM_ID, alarmIntent,
                PendingIntent.FLAG_CANCEL_CURRENT)

        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        // RTC does not wake the device up
        alarmManager.setRepeating(AlarmManager.RTC, calendar.timeInMillis, intervalMillis.toLong(), pendingIntent)
    }

    fun stopAlarm(context: Context, widgetId: Int, groupId: String) {
        val alarmIntent = Intent(context, FollowWidgetProvider::class.java)
        alarmIntent.action = C.ACTION.WIDGET_UPDATE
        alarmIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
        alarmIntent.putExtra(C.EXTRA.GROUP_ID, groupId)
        val pendingIntent = PendingIntent.getBroadcast(context, ALARM_ID, alarmIntent,
                PendingIntent.FLAG_CANCEL_CURRENT)

        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarmManager.cancel(pendingIntent)
    }
}
